### PR TITLE
Feat/depend

### PR DESCRIPTION
**Pull request by nikhedonia**
this package is a cli tool for dependency analysis for packages. 
It implements linting and automatic fixing of version numbers.
Furthermore it can print graphviz compatible description to console. 

Once merged we can add a linting and fix commands to our package.json files.
eg.:
`depend -e "packages/*/package.json" -s conservative --bail` which would result in an exit code of one if we have divergent package versions or a outdated times-component-package in use.

for quick fixing I'd suggest to use:
`depend -e "packages/*/package.json" -s majorityProgressive --fix`
which would automatically resolve all version number issues by the following rules:

if a dependency is a times-component package, take the latest version
  - if there are multiple versions used in the codebase use what the majority uses.
  - if it's a tie: use the latest version (progressive)

# More Examples

`depend -g "@times-components/*=>dextrose@" | fdp -Tsvg -o dextrose.svg`

dextrose

`depend -g "*/ad@ => *" | circo -Tgif -o ad.gif`

ad
 

`depend --hint --strategy majority`

```
packages/gestures/package.json
     dextrose: 1.3.2 -> 1.4.5
packages/slice/package.json
     @times-components/article: 0.32.7 -> 0.37.0
     dextrose: 1.4.0 -> 1.4.5
packages/tealium/package.json
     react-native-web: 0.1.14 -> 0.3.2
packages/utils/package.json
     @times-components/jest-configurator: 0.0.23 -> 0.2.0
```

`depend --list -e "packages/*/package.json"`

```
@times-components/watermark 0.3.4
2 prop-types 15.6.0
3 react-broadcast 0.5.2
4 @storybook/addon-actions 3.3.11
5 @storybook/react-native 3.3.11
6 @times-components/eslint-config-thetimes 0.1.3
7 dextrose 1.4.5 1.4.0 1.3.2
[...]
```

# QuickInstall

```
git ch feat/checkdep
npm pack packages/depend/
npm i -g times-components-depend-0.0.1.tgz
git ch - # back to work
depend -e "packages/*/package.json"--hint -s majorityProgressive # show all packages that need fixing
depend -e "packages/*/package.json" --fix -s majorityProgressive # fix what needs fixing  
```

# Limitations

currently ambitious version ranges are not supported.

### CODE REVIEW

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/50bd60f48fb3b35046eaa33ad21b77ad1770c464/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/f97510138fce2819a495d9dc3985aaa27666ac1c/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/c7d308fda7ac9e8e53ae86d208d27f3f22404781/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/2f2c2c30a3ad6ba618c36cc771449a3013fd2777/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/8a64ff9e9ec70ae356b3e9da190970873b3e4554/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/2ce2cfa24e9a22ee6253534884ee0968bd30c890/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/0e75714a39f786265b7ef242d5965d23fac30b3d/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/fe1ad8e709c14c56f31e3d32e03eaad2cfef2049/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/0b1d5bf9083598a408c9acf0cf4ff95d3208f940/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/4ffccab58b59b1534a07d4f89abb0ac8f1575d0f/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/93dd2d6b028e856b79c4a24b5421b16a61cccd7d/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/37904cc36a5c6d75eef7cc7b400d47d879fed23b/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/a2ecfa3d3db1eccfee00aeb3629a2eac0a03ff64/dextrose-test/index.html

File: packages/depend/README.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# depend/cli
+
+A tool for syncing version numbers in a monorepo.
+
+# Usage
+
+```
+depend -e "packages/*/package.json" [...options]
```
Comment (jimhunty) [inline packages/depend/README.md]: As it intends to always be a package dependency tool, cant we remove the need for `packages/*/package.json` ? Just seems like extra typing

File: packages/depend/README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+```
+
+# Options
+
+## --fix
+
+fixes the packages that are not in-sync
+
+## --graph, -g ["{filter}"]
```
Comment (jimhunty) [inline packages/depend/README.md]: Worth actually deploying these images somewhere and linking them into the PRs?

File: packages/depend/strategies.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import semver from "semver";
+
+function combine(s1, s2) {
+  return (a, b) => s1(a, b) || s2(a, b);
+}
+
+// TODO: support semver ranges
```
Comment (jimhunty) [inline packages/depend/strategies.js]: Worth adding as an issue or a Jira ticket

File: packages/depend/cli-options.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import optimist from "optimist";
```
Comment (jimhunty) [inline packages/depend/cli-options.js]: This packages is deprecated?

File: packages/depend/README.md (hunk 1 above)
Comment (nikhedonia) [reply packages/depend/README.md]: It actually defaults to `packages/*/package.json`; maybe it would be better to read this from the `lerna.json` file?.

However if you run it within `packages/ad` the expression will change.

I've also been using it a lot to analyse the node_modules folder to find out which dependency installs what.

File: packages/depend/cli-options.js (hunk 4 above)
Comment (nikhedonia) [reply packages/depend/cli-options.js]: true, couldn't get it working with yargs though...

File: packages/depend/README.md (hunk 2 above)
Comment (nikhedonia) [reply packages/depend/README.md]: It actually only creates a graphviz compatible output that can then be rendered to an image (see PR). Where would you guys suggest to upload those files ?

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/1bc12c451c2c619a3b8a71d4e96223b20d40ebc4/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/85d9c3ae12171aef51b5d6acf0400e76afe91bc3/dextrose-test/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/c5394c489f83befc3bebd4f00475e1bcd4bace6e/dextrose-test/index.html

File: packages/depend/README.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+## --expr, -e "{glob}"
+
+Defines the glob expression that will be used to search the filesystem for
+package.json files. By default it will use `"packages/*/package.json"`.
+
+## --fix
+
+Fixes the packages that are not in-sync
+
+## --graph, -g ["{filter}"]
```
Comment (craigbilner) [inline packages/depend/README.md]: not clear from the docs how you create the output? The filter section below doesn't appear related, example commands would be useful?

File: packages/depend/README.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+Prints a graphviz compatible output that can be converted to a graphic
+
+## --pick, -p {package}@{version}
+
+Adds `{package}@{version}` to the set of rules. Use this in conjunction with
```
Comment (craigbilner) [inline packages/depend/README.md]: this bit isn't too clear either, what should it do? Seems like it says prefer this version?

File: packages/depend/README.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+`*=>*` , ``, `*`
+
+everything
+
+## --bail
+
+exit with code 1 if packages are not in-sync
+
+## strategies [--strategy, -s]
```
Comment (craigbilner) [inline packages/depend/README.md]: Should there be a default strategy? This behaviour isn't obvious, if you don't provide a strategy then it'll only work on deps but with a strategy it works on dev deps?

File: packages/depend/README.md [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+pick the oldest used version
+
+### progressive
+
+pick the newest used version
+
+### majority
+
+use the version the is used by the majority of packages in your repo
```
Comment (craigbilner) [inline packages/depend/README.md]: typo: that

File: packages/depend/__tests__/strategies.test.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
+import {
+  conservative,
+  progressive,
+  majority,
+  majorityConservative,
+  majorityProgressive
+} from "../strategies";
+
+describe("depend strategy tests", () => {
```
Comment (craigbilner) [inline packages/depend/__tests__/strategies.test.js]: the suite names are a little odd? Should it just be `the depend strategy of ____ should ______`?

File: packages/depend/cli-options.js (hunk 4 above)
Comment (craigbilner) [reply packages/depend/cli-options.js]: can we try?

File: packages/depend/depend.js [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+function resolveConflicts(strategy, packages) {
+  const { name, version } = packages.sort(strategy)[0];
+  return { name, version };
+}
+
+export function suggestFix(packageJson = {}, rules = {}) {
+  const fixup = (dep = {}) =>
+    entries(dep)
+      .map(([key, value]) => [key, rules[key], value])
+      .filter(x => x[1])
```
Comment (craigbilner) [inline packages/depend/depend.js]: can you use destructuring with named args here and in other anonymous places as you have elsewhere?

File: packages/depend/depend.js [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+          version,
+          usedBy: [...flatReverseLookup[`${dependency}@${version}`]]
+        }))
+      ]
+    }))
+    .reduce(toObject, {});
+}
+
+// finds dependencies that don't match packages
+export function findWrongVersions(packages) {
```
Comment (craigbilner) [inline packages/depend/depend.js]: found this quite confusing because it just seems to be deps rather than wrong versions in general i.e. dev deps?

File: packages/depend/get-packages.js [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import glob from "glob";
+import { readJson } from "fs-extra";
+
+export default function getPackages(expr) {
+  return new Promise((done, reject) => {
```
Comment (craigbilner) [inline packages/depend/get-packages.js]: would traditionally be called `resolve`

File: packages/depend/package.json [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    "jest": "21.2.1",
+    "prettier": "1.8.2"
+  },
+  "dependencies": {
+    "babel-plugin-transform-runtime": "6.23.0",
+    "babel-polyfill": "6.26.0",
+    "babel-runtime": "6.26.0",
+    "chalk": "2.3.1",
+    "fs-extra": "5.0.0",
+    "glob": "^7.1.2",
```
Comment (craigbilner) [inline packages/depend/package.json]: we use exact throughout the repo

File: packages/depend/README.md (hunk 1 above)
Comment (nikhedonia) [reply packages/depend/README.md]: lerna support added

File: packages/depend/README.md (hunk 5 above)
Comment (nikhedonia) [reply packages/depend/README.md]: more examples added below

File: packages/depend/README.md [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+  "name": "bar",
+  "version": "0.0.2",
+  "dependencies": {
+    "react-native": "0.53.1",
+    "glob": "6.2.1"
+  }
+}
+```
+
+`depend -g` outputs all dependencies in `.dot` format:
```
Comment (craigbilner) [inline packages/depend/README.md]: graph has to take a pattern?

File: packages/depend/README.md [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+  "foo@0.0.1" -> "bar@0.0.1";
+}
+```
+
+### Rendering Svgs
+
+Depend itself does not have any rendering capabilities but `--graph` output can
+be directly piped into a [graphviz](https://www.graphviz.org/) renderer.
+
+`depend -g "@times-components/*=>dextrose@" | fdp -Tsvg -o dextrose.svg`
```
Comment (craigbilner) [inline packages/depend/README.md]: might want to mention you  need to have graphviz installed at this point?

File: packages/depend/README.md [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+
+#### Examples
+
+`=>react-native@0.50.*`
+
+everything that installs react-native ver0.50.\*
+
+`=>jest`
+
+everything that installs jest
```
Comment (craigbilner) [inline packages/depend/README.md]: this seems to output any package that contains `jest` in it?

File: packages/depend/__tests__/apply-fix.test.js [hunk 17]
```diff
@@ -0,0 +1,54 @@
+import { applyPatch } from "../depend";
+
+describe("depend apply-patch tests", () => {
```
Comment (craigbilner) [inline packages/depend/__tests__/apply-fix.test.js]: ideally these should read as sentences

File: packages/depend/package.json [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+  "version": "0.0.1",
+  "description": "dependency analyzer and fixer",
+  "main": "depend",
+  "scripts": {
+    "build": "babel . -d dist --ignore dist,node_modules,coverage,__tests__ && cp package.json dist/package.json",
+    "prepublish": "npm run build",
+    "fmt": "prettier --write '**/*.*'",
+    "prettier:diff": "prettier --list-different '**/*.*'",
+    "lint": "eslint . && npm run prettier:diff",
+    "test": "jest --bail --ci --coverage"
```
Comment (craigbilner) [inline packages/depend/package.json]: we don't use these flags at the package level anymore

File: packages/depend/depend.js [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+  return todo.map(({ path, packageJson, patch }) => [
+    path,
+    [
+      ...entries(patch.dependencies)
+        .map(([name, version]) => [
+          name,
+          packageJson.dependencies[name],
+          version
+        ])
+        .filter(x => x[1]),
```
Comment (craigbilner) [inline packages/depend/depend.js]: still some cryptic refs like this that should be refactored
